src/gf.c

```diff
@@ -1541,28 +1541,203 @@ static jl_value_t *verify_type(jl_value_t *v)
+              hash-consed, so pointer comparison should work.
+            */
+            return 0;
+        }
+    }
+    return 1;
+}
+
+//STATIC_INLINE jl_typemap_entry_t *sig2_match_fast(jl_value_t **args, jl_typemap_entry_t *entry1, jl_typemap_entry_t *entry2, size_t n)
+//{
+//    // NOTE: This function is a huge performance hot spot!!
+//    jl_value_t **sig1 = jl_svec_data(entry1->sig->parameters);
+//    jl_value_t **sig2 = jl_svec_data(entry2->sig->parameters);
+//    size_t i;
+//    for (i = 0; i < n; i++) {
+//        jl_value_t *decl1 = sig1[i];
+//        jl_value_t *decl2 = sig2[i];
+//        jl_value_t *a = args[i];
+//        jl_value_t *ta = jl_typeof(a);
+//        if (ta != decl1) {
+//            /*
+//              we are only matching concrete types here, and those types are
+//              hash-consed, so pointer comparison should work.
+//            */
+//            return sig_match_fast(args, sig2, i, n) ? entry2 : NULL;
+//        }
+//        if (ta != decl2) {
+//            /*
+//              we are only matching concrete types here, and those types are
+//              hash-consed, so pointer comparison should work.
+//            */
+//            return sig_match_fast(args, sig1, i, n) ? entry1 : NULL;
+//        }
+//    }
+//    return entry1;
+//}
+
+jl_typemap_entry_t *call_cache[N_CALL_CACHE];
+static uint8_t pick_which[N_CALL_CACHE];
+#ifdef JL_GF_PROFILE
+void call_cache_stats() {
+    int pick_which_stat[4] = {0, 0, 0, 0};
+    int i, count = 0;
+    for (i = 0; i < N_CALL_CACHE; i++) {
+        if (call_cache[i])
+            count++;
+    }
+    for (i = 0; i < N_CALL_CACHE; i++) {
+        ++pick_which_stat[pick_which[i] & 3];
+    }
+    jl_safe_printf("cache occupied: %d / %d; pick_which stats: {%d, %d, %d, %d}\n",
+            count, N_CALL_CACHE,
+            pick_which_stat[0], pick_which_stat[1], pick_which_stat[2], pick_which_stat[3]);
+}
+#endif
+
 JL_DLLEXPORT jl_value_t *jl_apply_generic(jl_value_t **args, uint32_t nargs)
 {
-    jl_value_t *F = args[0];
-    jl_methtable_t *mt = jl_gf_mtable(F);
 #ifdef JL_GF_PROFILE
     mt->ncalls++;
 #endif
 #ifdef JL_TRACE
     int traceen = trace_en; //&& ((char*)&mt < jl_stack_hi-6000000);
     if (traceen)
-        show_call(F, &args[1], nargs-1);
+        show_call(args[0], &args[1], nargs-1);
 #endif
+
     /*
       search order:
+      check associative hash based on callsite address for leafsig match
       look at concrete signatures
       if there is an exact match, return it
       otherwise look for a matching generic signature
       if no concrete or generic match, raise error
       if no generic match, use the concrete one even if inexact
       otherwise instantiate the generic method and use it
     */
-    jl_typemap_entry_t *entry = jl_typemap_assoc_exact(mt->cache, args, nargs, jl_cachearg_offset(mt));
+    int callsite = int32hash((uintptr_t)__builtin_return_address(0));
```

Comment:
I'm currently assuming it'll be available in some form on all compilers.

E.g, for MSVC, we'll need `#define __builtin_return_address(n) _ReturnAddress()`
https://msdn.microsoft.com/en-us/library/64ez38eh%28v=vs.140%29.aspx